src/server/player_sao.cpp

```diff
@@ -457,20 +457,28 @@ u16 PlayerSAO::punch(v3f dir,
 
 void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason)
 {
-	s32 oldhp = m_hp;
+	if (hp == (s32)m_hp)
+		return; // Nothing to do
 
-	hp = rangelim(hp, 0, m_prop.hp_max);
+	if (m_hp <= 0 && hp < (s32)m_hp)
+		return; // Cannot take more damage
 
-	if (oldhp != hp) {
-		s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - oldhp, reason);
+	if (m_hp >= m_prop.hp_max && hp > (s32)m_hp)
+		return; // Cannot heal more
```

Comment:
Wouldn't it be better to run the callback in these two cases so mods have the fully flexibility?

Reply:
This was added on purpose due to https://github.com/minetest/minetest/pull/8462#issuecomment-482242096

Reply:
Hm, i'm wondering if this doesn't exclude some possible usecases.
If a mod wants to e.g. remember HP that were healed "too much", does it have the opportunity to do so if on_player_hpchange is not called?

Reply:
Modders could only detect it the first time when the new health is >= the maximal HP value. I don't see any use-case for this yet, but I removed this check in case modders need it.